src/hooks.ts

```diff
@@ -201,38 +262,60 @@ export const useExperiment: UseExperiment = (experimentKey, options = {}, overri
   if (!optimizely) {
     throw new Error('optimizely prop must be supplied via a parent <OptimizelyProvider>');
   }
 
-  // Helper function to return the current values for isEnabled and variables.
-  const getCurrentValues = useCallback<() => FeatureDecisionValues>(
+  const overrideAttrs = useCompareAttrsMemoize(overrides.overrideAttributes);
+  const getCurrentDecision: () => FeatureDecisionValues = useCallback(
     () => ({
-      isEnabled: optimizely.isFeatureEnabled(featureKey, overrides.overrideUserId, overrides.overrideAttributes),
-      variables: optimizely.getFeatureVariables(featureKey, overrides.overrideUserId, overrides.overrideAttributes),
+      isEnabled: optimizely.isFeatureEnabled(featureKey, overrides.overrideUserId, overrideAttrs),
+      variables: optimizely.getFeatureVariables(featureKey, overrides.overrideUserId, overrideAttrs),
     }),
-    [featureKey, overrides]
+    [optimizely, featureKey, overrides.overrideUserId, overrideAttrs]
   );
 
-  // Set the initial state immediately serverSide
-  const [state, setState] = useState<UseFeatureState>(() => {
-    const initialState = {
-      isEnabled: false,
-      variables: {},
-      clientReady: isServerSide ? true : false,
-      didTimeout: false,
-    };
-    if (isServerSide) {
-      return { ...initialState, ...getCurrentValues() };
+  const isClientReady = isServerSide || optimizely.isReady();
+  const decisionStateAndSetter = useState<FeatureDecisionValues>(() => {
+    if (isClientReady) {
+      return getCurrentDecision();
     }
-    return initialState;
+    return { isEnabled: false, variables: {} };
   });
+  let decisionState = decisionStateAndSetter[0];
+  const setDecisionState = decisionStateAndSetter[1];
+  // Decision state is derived from entityKey and overrides arguments.
+  // Track the previous value of those arguments, and update state when they change.
+  // This is an instance of the derived state pattern recommended here:
+  // https://reactjs.org/docs/hooks-faq.html#how-do-i-implement-getderivedstatefromprops
+  const currentDecisionInputs: DecisionInputs = {
+    entityKey: featureKey,
+    overrideUserId: overrides.overrideUserId,
+    overrideAttributes: overrides.overrideAttributes,
+  };
+  const [prevDecisionInputs, setPrevDecisionInputs] = useState<DecisionInputs>(currentDecisionInputs);
+  if (!areDecisionInputsEqual(prevDecisionInputs, currentDecisionInputs)) {
+    setPrevDecisionInputs(currentDecisionInputs);
+    decisionState = getCurrentDecision();
+    setDecisionState(decisionState);
+  }
 
-  useEffect(
-    initializeWhenClientReadyFn(HookType.FEATURE, featureKey, optimizely, options, timeout, setState, getCurrentValues),
-    [optimizely]
-  );
+  useEffect(() => {
+    if (!isClientReady || options.autoUpdate) {
+      return setupAutoUpdateListeners(optimizely, HookType.FEATURE, featureKey, hooksLogger, () => {
+        setDecisionState(getCurrentDecision());
+      });
+    }
+    return (): void => {};
```

Comment:
Will we have a memory leak issue on unmount? Do we need to remove auto update listeners here?

Reply:
I think it's ok because if control reaches this line, no auto update listeners were added.

Reply:
Oh, rightttt, my misread on this. That's clear to me now.